server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java

```diff
@@ -1135,6 +1161,116 @@ public synchronized void updateGlobalCheckpointForShard(final String allocationI
+     *
+     * This method will also stop timers for each shard and compute replication lag metrics.
+     *
+     * @param allocationId     the allocation ID to update the global checkpoint for
+     * @param visibleCheckpoint the visible checkpoint
+     */
+    public synchronized void updateVisibleCheckpointForShard(final String allocationId, final ReplicationCheckpoint visibleCheckpoint) {
+        assert indexSettings.isSegRepEnabled();
+        assert primaryMode;
+        assert handoffInProgress == false;
+        assert invariant();
+        final CheckpointState cps = checkpoints.get(allocationId);
+        assert !this.shardAllocationId.equals(allocationId) && cps != null;
+        if (cps.checkpointTimers.isEmpty() == false) {
+            // stop any timers for checkpoints up to the received cp and remove from cps.checkpointTimers.
+            // Compute the max lag from the set of completed timers.
+            final AtomicLong lastFinished = new AtomicLong(0L);
+            cps.checkpointTimers.entrySet().removeIf((entry) -> {
+                boolean result = visibleCheckpoint.equals(entry.getKey()) || visibleCheckpoint.isAheadOf(entry.getKey());
+                if (result) {
+                    final ReplicationTimer timer = entry.getValue();
+                    timer.stop();
+                    lastFinished.set(Math.max(lastFinished.get(), timer.time()));
+                }
+                return result;
+            });
+            cps.lastCompletedReplicationLag = lastFinished.get();
+        }
+        logger.trace(
+            () -> new ParameterizedMessage(
+                "updated local knowledge for [{}] on the primary of the visible checkpoint from [{}] to [{}], active timers {}",
+                allocationId,
+                cps.visibleReplicationCheckpoint,
+                visibleCheckpoint,
+                cps.checkpointTimers.keySet()
+            )
+        );
+        cps.visibleReplicationCheckpoint = visibleCheckpoint;
+        assert invariant();
+    }
+
+    /**
+     * After a new checkpoint is published, start a timer for each replica to the checkpoint.
+     * @param checkpoint {@link ReplicationCheckpoint}
+     */
+    public synchronized void setLatestReplicationCheckpoint(ReplicationCheckpoint checkpoint) {
+        assert indexSettings.isSegRepEnabled();
+        assert primaryMode;
+        assert handoffInProgress == false;
+        if (checkpoint.equals(lastPublishedReplicationCheckpoint) == false) {
+            this.lastPublishedReplicationCheckpoint = checkpoint;
+            for (Map.Entry<String, CheckpointState> entry : checkpoints.entrySet()) {
+                if (entry.getKey().equals(this.shardAllocationId) == false) {
+                    final CheckpointState cps = entry.getValue();
+                    if (cps.inSync) {
+                        cps.checkpointTimers.computeIfAbsent(checkpoint, ignored -> {
```

Comment:
I think we also need to accomodate below:
1. Handle cluster manager updates for shard failures, index deletions etc and remove required entries. This may need us to use a different key for map here (shardId/allocation id) ?
2. Audit timers from time to time and remove stale entries. This is to account for local clean up failures and keeps a tap on consumed memory.

Reply:
Good questions.

1. This is actually why I chose to reuse ReplicationTracker & CheckpointState vs storing these stats in a new component.  This component is already alerted to any cluster state update through the `updateFromClusterManager` method and removes entries from its checkpoints map [here](https://github.com/opensearch-project/OpenSearch/blob/3a135e926f703e8e17d785c0ad15c88763e6d23b/server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java#L1371).  This will remove the entry if the allocationId falls out of sync (shard failure/index deletion etc) & is not initializing.

2.  Could you elaborate a bit on "local clean up failures?"  If the shard is no longer in sync it would get removed, so I'm not sure what other case this would cover.  We will also be failing the shard if it is over a certain count of checkpoints & time behind forcing its removal.  I'm not opposed to however adding a limit on the count of timers a shard could have correlating to these limits, so that we aren't having to wait for the failure event.

Reply:
I played around with storing only a single ReplicationTimer here with CheckpointState & storing a queue of published checkpoints instead of the map.  The problem here is we ened to start a timer with each published checkpoint, or our metric is incorrectly shorter than it should be as its the lag between events vs the lag behind the primary.  We can reduce memory footprint here a bit and store the infos version as our key instead of the entire checkpoint, given this data is wiped on failover/handoff.